Mage/src/main/java/mage/target/common/TargetCardAndOrCardTypeInLibrary.java

```diff
@@ -0,0 +1,74 @@
+package mage.target.common;
+
+import mage.abilities.Ability;
+import mage.abilities.dynamicvalue.common.CardTypeAssignment;
+import mage.cards.Card;
+import mage.cards.Cards;
+import mage.cards.CardsImpl;
+import mage.constants.CardType;
+import mage.filter.FilterCard;
+import mage.filter.predicate.Predicates;
+import mage.game.Game;
+import mage.util.CardUtil;
+
+import java.util.UUID;
+
+/**
+ * @author xenohedron
+ */
+public class TargetCardAndOrCardTypeInLibrary extends TargetCardInLibrary {
+
+    private static FilterCard makeFilter(CardType firstType, CardType secondType) {
+        FilterCard filter = new FilterCard(
+                CardUtil.addArticle(CardUtil.getTextWithFirstCharLowerCase(firstType.toString()))
+                + " card and/or "
+                + CardUtil.addArticle(CardUtil.getTextWithFirstCharLowerCase(secondType.toString()))
+                + " card");
+        filter.add(Predicates.or(
+                firstType.getPredicate(), secondType.getPredicate()
+        ));
+        return filter;
+    }
+
+    private final CardTypeAssignment assignment;
+
+    /**
+     * a [firstType] card and/or a [secondType] card
+     */
+    public TargetCardAndOrCardTypeInLibrary(CardType firstType, CardType secondType) {
+        super(0, 2, makeFilter(firstType, secondType));
+        this.assignment = new CardTypeAssignment(firstType, secondType);
+    }
+
+    protected TargetCardAndOrCardTypeInLibrary(final TargetCardAndOrCardTypeInLibrary target) {
+        super(target);
+        this.assignment = target.assignment;
+    }
+
+    @Override
+    public TargetCardAndOrCardTypeInLibrary copy() {
+        return new TargetCardAndOrCardTypeInLibrary(this);
+    }
+
+    @Override
+    public boolean canTarget(UUID playerId, UUID id, Ability source, Game game) {
+        if (!super.canTarget(playerId, id, source, game)) {
+            return false;
+        }
+        Card card = game.getCard(id);
+        if (card == null) {
+            return false;
+        }
+        if (this.getTargets().isEmpty()) {
+            return true;
+        }
+        Cards cards = new CardsImpl(this.getTargets());
+        cards.add(card);
+        return assignment.getRoleCount(cards, game) >= cards.size();
```

Comment:
For info: some AI or playable code can depends on `Target::possibleTargets` implementation too. Well, in most use cases it will be fine cause it checking each possible target by `canTarget`, but some AI or human code can miss of `canTarget` calls). Smells bugs for such use case: AI can targeting invalid targets, e.g. ignore some restrictions (looks like a recent bug from #11491).

Reply:
So if you use additional filters in `Target` class, additional predicate fields in `Filter` class or additional logic with outside data in `canTarget` then pay attention to possible targeting issues.

Reply:
From TargetCard card:
> // TODO: check all class targets, if it override canTarget then make sure it override ALL 3 METHODS with canTarget and possibleTargets (method with cards doesn't need)

![shot_231201_075725](https://github.com/magefree/mage/assets/8344157/f9bb93c7-745d-4345-ad5a-1bbca5831548)